// Copyright 2018 Intel Corporation
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

option java_multiple_files = true;
option java_package = "pdo.sawtooth.protobuf";

message PdoContractEnclaveInfo {
    // Encoded public key corresponding to private key used the PDO contract
    // enclave to sign contract updates.
    // It is also serves an id to generate address for this entry
    string verifying_key = 1;

    // RSA key used to encrypt messages to the enclave
    string encryption_key = 2;

    // Public key of the signer who submitted "register" transaction
    // for the contract enclave
    // "Revoke" and "update" transaction signuatures must be verified
    // using this key
    string owner_id = 3;

    // Block identifier copied from the most recent registration
    // of this enclave used to ensure that the enclave
    // has been re-registered recently enough
    string last_registration_block_context = 4;

    // This is "register" transaction signature hash. It can be used
    // to get all registration transaction details from the ledger.
    string registration_transaction_id = 5;

    // Information that can be used internally to verify the validity of
    // the registration information stored as an opaque buffer.
    // This is a serialized json object containing three fields:
    // - 'verification_report': the body of the IAS verification response
    // -'certificates': a list with the certificate path (as PEM strings) of the IAS signing authority, with the signer first and the CA second. this corresponds to the x-iasreport-signing-certificate header of the IAS response
    // - 'signature': the signature on the report. This is the x-iasreport-signature header of the IAS response
    string proof_data = 6;
}


message PdoContractEnclaveTransaction {
    // The action that the transaction processor will take.
    // Currently this is only “register”, but could include other actions
    // in the futures such as “revoke”, "update", or "refresh"
    string verb = 1;

    // Encoded public key corresponding to private key used the PDO contract
    // enclave to sign contract updates.
    // It is also serves an id to generate address for this entry
    string verifying_key = 2;

    // Transaction details specific to the action. Currently only details
    // for "register" action are defined - PdoContractEnclaveRegister
    bytes transaction_details = 3;
}


message PdoContractEnclaveRegister {
    // Reserved for future use; currently ignored
    string organizational_info = 1;

    // RSA key used to encrypt messages to the enclave
    string encryption_key = 2;

    // Information that can be used internally to verify the validity of
    // the registration information stored as an opaque buffer.
    // See above for content encoding
    string proof_data = 3;

    // EPID pseudonym that identifies the host on which
    // the enclave resides, knowing that two enclaves reside on the
    // same host enables provisioning policies that distribute across hosts
    string enclave_persistent_id = 4;

    // This is the identity of a block in the chain.
    // This prevents old registrations from being included.
    // No known threats from the old registrations at this point
    string registration_block_context = 5;
}

message PdoContractEnclaveUpdate {
    // This is the identity of a block in the chain.
    string registration_block_context = 1;
}

